### ISSUE TITLE
`(require openssl)` on Linux requires `libcrypto.so` hence openssl-devel

### ISSUE THREAD

**Issue by Blaisorblade** [racket/racket#1096]
As @rmculpepper & others [figured out elsewhere](https://github.com/mbutterick/pollen/issues/52), currently (on 6.2.1) `(require openssl`) can fail on Linux default installs, because [`openssl` looks for](https://github.com/racket/racket/blob/master/racket/collects/openssl/libcrypto.rkt) `libcrypto.so` and `libcrypto.so.SomeSuffixes`, while only `libcrypto.so.SomeOtherSuffix` is available.

What's worse, there's a bug in the loading code such that the error message is `ssl-connect: requested protocol not supported`, while all the right information is available to Racket:
```
> (require openssl)
> ssl-available?
#f
> ssl-load-fail-reason
"ffi-lib: couldn't open \"libcrypto.so\" (libcrypto.so: cannot open shared object file: No such file or directory)"
```

It seems that one should:
- Update the error message for failed loading.
- Then, update the version list
- Finally, if possible, figure out how to avoid needing one.

If that matters, this issue affected one first-semester student on an HtDP-based course, where such a bug is especially unfortunate. At least we found the above thread with a diagnosis.

**Comment by rmculpepper** [comment]
Commits 49dc0625 and 3e2b916f attempt to address your first and second recommendations above, and they will be included in Racket 6.3, to be released soon.

**Comment by juanfra684** [comment]
FYI, OpenBSD doesn't use "version-less" libraries and openssl/libcrypto works fine. In fact, the actual version of the library is `libcrypto.so.36.1`.

**Comment by bremner** [comment]
[openssl.txt](https://github.com/racket/racket/files/818697/openssl.txt)

This patch, along with installing libssl1.1, seems to fix the problem for me.

I agree that a less hardcoded approach would be nice; normally on Debian these kinds of questions are resolved at link time, but of course that doesn't work for dlopened libraries.

**Comment by Blaisorblade** [comment]
Using dlopen isn't such a hard obstacle:
If you did run a ./configure script or equivalent, say while building a Racket .deb package (in Debian/Ubuntu PPAs), you could probably run pkgconfig and/or readlink to figure out the right version of dependencies. Part of the trick is that while building packages you can more reasonably depend on development packages.
The harder part is to figure out how to deal with non-.deb-based installs.

Meanwhile, Racket Debian packages could have a soft dependency on libssl-dev (probably recommends and not suggests). That *is* still a hack, but a more robust one on most fronts (except disk usage). And maybe dev. packages on a server increase security risks?

**Comment by bremner** [comment]
And after I figure out the right version, then what? racket still needs to be patched to load it.
I'm not very happy about the idea of recommending libssl-dev; this means when there's an SONAME bump, a new version will be brought in that hasn't been tested with racket. We already say that the bump to 1.1 required patching racket (not just the library loading, but changing the API used).

**Comment by Blaisorblade** [comment]
You're right, I ignored API changes as I wasn't aware of them. Those certainly need be handled by hand, as for other clients. I was just answering to:
> I agree that a less hardcoded approach would be nice; normally on Debian these kinds of questions are resolved at link time, but of course that doesn't work for dlopened libraries.

And there you didn't refer to API changes, did you?

IIUC, that means C code links against `-lssl` with `libssl-dev` installed, producing a binary linking against `libssl.so.N`. My idea was to emulate the same result by learning `N` and patching Racket code à-la configure. Say, for version `$SSL_VERSION`, have in the `.deb` package build script something like:

```
sed -i "s/@@SSL_LIBRARY_VERSION@@/$SSL_VERSION/" .../$PathToRacket/collects/openssl/libversion.rkt
```

That's equivalent to what's done usually at link time, and avoids the hardcoded version table: https://github.com/racket/racket/blob/1cbd75436b3a10090e7ff9094e64caef4876305f/racket/collects/openssl/libcrypto.rkt#L12-L53.

Again, nothing saves you with API updates such as the one for 1.1 (found it in b0c55b73945d8dab52f8ff86df69dd6cbb0147b4). But that's no argument for the status quo.

> I'm not very happy about the idea of recommending libssl-dev; this means when there's an SONAME bump, a new version will be brought in that hasn't been tested with racket.

I agree that would be a consequence. It's not obvious to me that's worse.

By that argument, maybe Racket should stop ever loading `libssl.so`. But I'm proposing to change the default. What's the danger though? This depends on OpenSSL compatibility policies:
- If used symbols are removed, failing when loading the library seems no different. If symbols are loaded lazily, code might fail later which would be more annoying.
- If used symbols change behavior arbitrarily, the worst case is that previously correct API usage misbehaves (detectable by testing) or becomes vulnerable (not detectable by testing). That would be actually bad, and warrants never loading `libssl.so`.

**Comment by bremner** [comment]
"Paolo G. Giarrusso" <notifications@github.com> writes:

>
>> I'm not very happy about the idea of recommending libssl-dev; this means when there's an SONAME bump, a new version will be brought in that hasn't been tested with racket.
>
> I agree that would be a consequence. It's not obvious to me that's worse.
>
> By that argument, maybe Racket should stop ever loading `libssl.so`.

From my point of view as a distro packager, loading libssl.so is indeed
a bad idea.
